rfcs/012 - Either, Option, and Result Destructuring.md

```diff
@@ -0,0 +1,110 @@
+if myResult.isOk() {
+	let okVal = myResult.getOr(-1);
+	print("should be 5: " + okVal);
+} else {
+	let errVal = myResult.getErr("expected error, but could not retrieve Error value");
+	print("unreachable, but I know that errVal is NOT 'expected error, but could not retrieve Error value', unless that's what myResult is");
+}
+```
+
+Note that, in both branches of the above `if` statement, the status (`ok` or `err`) of the `Result` is already known, but the user must then call either `getOr` or `getErr` in order to retrieve the expected value.
+(Note that `getOrDefault` is not yet implemented, but only makes this process marginally easier.)
+Instead, it would be much easier for the user (and potentially even more efficient, although marginally) if the variable could be immediately used as if it were destructured.
+Since `isOk` already guarantees that the `Result` is `ok` (when it returns `true`, otherwise the guarantee is that it's `err`), it would make sense to be able to write the above code like this:
+```
+let myResult = ok(5);
+if myResult.isOk() {
+	print("should be 5: " + myResult); // 1
+} else {
+	print("unreachable, but if myResult is an error, then this would print that error: " + myResult); // 2
+}
+```
+
+In the above code, `myResult` was already destructured into its `ok` value at location 1
+(in other words, at location 1, `myResult` is of type `int64`, not `Result<int64>`).
+This allows the user to quickly and easily perform operations on that inner value without the additional step mentioned above.
+Note that, if myResult were an `err` instead, it would work the exact same way, with the `Error` taking the place of `myResult`.
+
+This should also be extended to `Option<T>` and `Either<T, U>` as well:
+```
+let myOption = some(true);
+if myOption.isSome() {
+	print("here ya go: " + myOption); // 1
+} else {
+	print("I've got nothing for ya"); // 2
+}
+
+let myEither = main("hello, world!"); // 3
+if myEither.isMain() {
+	print("main rock: " + myEither); // 4
+} else {
+	print("alt rock: " + myEither); // 5
+}
+```
+
+In location 1, `myOption` should be of type `bool`.
+Meanwhile, in location 2, it might make sense to undefine the `myOption` var - this is fairly ambiguous and needs more discussion.
+Note that in location 3, `myEither` would technically be assigned the type `Either<string, void>`, but imagine it somehow assigned to a stringable value, say an `int64`.
+In location 4, `myEither` should be of type `string`, and location 5 would be `int64` or such.
+
+The implementation should be relatively simple: when building the `amm` representation of the program, if the compiler detects any `if` statements (and other constructs that should have this feature too - if deemed worthy) with this pattern, then it should reassign the value at the beginning of the block.
+Preferably, this would be done with the `getR` opcode, which would allow for the value to be destructured from the `Result` efficiently.
+
+### Alternatives Considered
```

Comment:
We should also raise the consideration about complex conditionals that still prove that the value must be true and how we're avoiding that for now due to the far higher complexity in the compiler necessary to support that.

I would also like to add an alternative considered being early return from one of these conditional blocks implying the opposite state for the remainder of the outer scope as something considered but kept out not because it would be hard (we already isolate the remainder of the scope when there's an early return to make sure the dependency graph is executed correctly) but because it would cause hilarity when refactoring code.

Reply:
Oh, one last alternative. We could have a bare `get()` method that is only allowed within the conditional scope that has this conditional check and the user can unwrap safely without providing a default value, but it's really weird to have a method that can't always be called on the variable type it's designed for.